File: code/cpp/ch13/practice/practice_05_combination_sum.cpp
/*
 * Practice 5: Combination Sum
 * ====================================
 * Chapter 13: Bronze Battle Plan — Complete Search & Simulation
 *
 * PROBLEM:
 *   Find all unique combinations of candidates that sum to target.
 *   Each number may be reused unlimited times.
 *
 * EXAMPLES:
 *   solve({2,3,6,7}, 7) -> {{2,2,3},{7}}
 *
 * CONSTRAINTS:
 *   - 1 <= candidates.size() <= 10
 *   - 1 <= candidates[i] <= 30
 *   - 1 <= target <= 30
 */

#include <algorithm>
#include <iostream>
#include <vector>
using namespace std;

vector<vector<int>> solve(vector<int> candidates, int target) {
    // TODO: Replace this with your solution
    return {};
}

// -- Do not change anything below this line --------------------------
int main() {
    int n, target;
    cin >> n;
    vector<int> candidates(n);
    for (int i = 0; i < n; i++) cin >> candidates[i];
    cin >> target;
    vector<vector<int>> result = solve(candidates, target);
    for (auto& combo : result) {
        for (int i = 0; i < (int)combo.size(); i++) {
            if (i > 0) cout << " ";
            cout << combo[i];
        }
        cout << endl;
    }
    return 0;
}

File: code/python/ch13/practice/practice_05_combination_sum.py
"""
Practice 5: Combination Sum
==============================
Chapter 13: Bronze Battle Plan — Putting It All Together

PROBLEM
-------
Given a list of distinct positive integers (candidates) and a target
integer, find all unique combinations where the candidates sum to the
target. Each candidate may be used unlimited times. Return combinations
in sorted order.

INPUT FORMAT
------------
First line: space-separated integers (candidates).
Second line: a single integer (target).

OUTPUT FORMAT
-------------
Each combination on its own line as a list.

CONSTRAINTS
-----------
- 1 <= len(candidates) <= 30
- 1 <= candidates[i] <= 200
- 1 <= target <= 500
- All candidates are distinct

EXAMPLES
--------
Input:
  2 3 6 7
  7
Output:
  [2, 2, 3]
  [7]

Input:
  2 3 5
  8
Output:
  [2, 2, 2, 2]
  [2, 3, 3]
  [3, 5]

INSTRUCTIONS
------------
Replace the `pass` in the solve() function with your solution.
The main block at the bottom handles input/output — don't change it.
"""


def solve(candidates: list[int], target: int) -> list[list[int]]:
    """Return all unique combinations summing to target."""
    pass  # TODO: Replace this with your solution



# ── Do not change anything below this line ──────────────────────────
if __name__ == "__main__":
    candidates = list(map(int, input().split()))
    target = int(input())
    result = solve(candidates, target)
    for combo in result:
        print(combo)

